### ISSUE TITLE
ConnectionRefuseError

### ISSUE THREAD

**Issue by negodfre** [t-wolfeadam/Alpyne#4]
I was interested in using Alpyne; however, I was having difficulty getting it to run on my machine.

Java Version "17.0.1"
Javac Version "17.0.1"

I was testing the "Stock Management Game".
The "simple run - single.py", but I keep getting a "Connection Refuse Error". 

I proceeded to test just the part that fails which is the following:
`client = AlpyneClient(r"Exported\StockManagementGame\model.jar", blocking=True, verbose=True)`

Does anyone happen to know what might be the issue?
I wasn't sure if it was a security issue or something related to pathing, variables, etc.

**Error / Traceback:**
`---------------------------------------------------------------------------
ConnectionRefusedError                    Traceback (most recent call last)
C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\urllib\request.py in do_open(self, http_class, req, **http_conn_args)
   1345             try:
-> 1346                 h.request(req.get_method(), req.selector, req.data, headers,
   1347                           encode_chunked=req.has_header('Transfer-encoding'))

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\http\client.py in request(self, method, url, body, headers, encode_chunked)
   1284         """Send a complete request to the server."""
-> 1285         self._send_request(method, url, body, headers, encode_chunked)
   1286 

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\http\client.py in _send_request(self, method, url, body, headers, encode_chunked)
   1330             body = _encode(body, 'body')
-> 1331         self.endheaders(body, encode_chunked=encode_chunked)
   1332 

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\http\client.py in endheaders(self, message_body, encode_chunked)
   1279             raise CannotSendHeader()
-> 1280         self._send_output(message_body, encode_chunked=encode_chunked)
   1281 

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\http\client.py in _send_output(self, message_body, encode_chunked)
   1039         del self._buffer[:]
-> 1040         self.send(msg)
   1041 

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\http\client.py in send(self, data)
    979             if self.auto_open:
--> 980                 self.connect()
    981             else:

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\http\client.py in connect(self)
    945         """Connect to the host and port specified in __init__."""
--> 946         self.sock = self._create_connection(
    947             (self.host,self.port), self.timeout, self.source_address)

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\socket.py in create_connection(address, timeout, source_address)
    843         try:
--> 844             raise err
    845         finally:

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\socket.py in create_connection(address, timeout, source_address)
    831                 sock.bind(source_address)
--> 832             sock.connect(sa)
    833             # Break explicitly a reference cycle

ConnectionRefusedError: [WinError 10061] No connection could be made because the target machine actively refused it

During handling of the above exception, another exception occurred:

URLError                                  Traceback (most recent call last)
~\.environments\fastai\lib\site-packages\alpyne\client\alpyne_client.py in __init__(self, model_loc, blocking, port, verbose)
     48             self._http_client = HttpClient(f"http://127.0.0.1:{port}")
---> 49             _, _, output = self._http_client.get("/versions/number/0")
     50             self.version = ModelVersion.from_json(output)

~\.environments\fastai\lib\site-packages\alpyne\client\http_client.py in get(self, url)
     56     def get(self, url: str) -> RequestResponse:
---> 57         return self.api_request(url, method="GET")
     58 

~\.environments\fastai\lib\site-packages\alpyne\client\http_client.py in api_request(self, url, method, body)
     37             request.add_header("Content-Type", "application/json")
---> 38             response = urllib.request.urlopen(request, timeout=3)
     39             if response.getcode() >= 400:  # TODO model errors should warn user and mark episode as halted

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\urllib\request.py in urlopen(url, data, timeout, cafile, capath, cadefault, context)
    213         opener = _opener
--> 214     return opener.open(url, data, timeout)
    215 

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\urllib\request.py in open(self, fullurl, data, timeout)
    516         sys.audit('urllib.Request', req.full_url, req.data, req.headers, req.get_method())
--> 517         response = self._open(req, data)
    518 

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\urllib\request.py in _open(self, req, data)
    533         protocol = req.type
--> 534         result = self._call_chain(self.handle_open, protocol, protocol +
    535                                   '_open', req)

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\urllib\request.py in _call_chain(self, chain, kind, meth_name, *args)
    493             func = getattr(handler, meth_name)
--> 494             result = func(*args)
    495             if result is not None:

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\urllib\request.py in http_open(self, req)
   1374     def http_open(self, req):
-> 1375         return self.do_open(http.client.HTTPConnection, req)
   1376 

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2288.0_x64__qbz5n2kfra8p0\lib\urllib\request.py in do_open(self, http_class, req, **http_conn_args)
   1348             except OSError as err: # timeout error
-> 1349                 raise URLError(err)
   1350             r = h.getresponse()

URLError: <urlopen error [WinError 10061] No connection could be made because the target machine actively refused it>

During handling of the above exception, another exception occurred:

ModelError                                Traceback (most recent call last)
~\AppData\Local\Temp/ipykernel_10116/3414797435.py in <module>
----> 1 client = AlpyneClient(r"Exported\StockManagementGame\model.jar", blocking=True, verbose=True)

~\.environments\fastai\lib\site-packages\alpyne\client\alpyne_client.py in __init__(self, model_loc, blocking, port, verbose)
     50             self.version = ModelVersion.from_json(output)
     51         except:
---> 52             raise ModelError(f"Failed to properly start the app. Error from process: {self._proc.stderr.readline().decode()}")
     53 
     54     def _validate_java(self):

ModelError: ModelError[path=, status=Failed to properly start the app. Error from process: Error: A JNI error has occurred, please check your installation and try again
, error=None, msg='']`

**Comment by t-wolfeadam** [comment]
Hello! From the last line it seems there might be a problem with your Java installation. I'd check via `java -version` and `javac -version` that they're both pointing to the same version, and one that's 11+. 

If that checks out, please try re-running your script with verbose=True and upload the alpyne.log file here - assuming you're on the latest version (0.1.2), it will be in your script directory.

**Comment by negodfre** [comment]
Hello!

I installed the newer versions of Java.
When I check my versioning, it shows "17.0.1" for both Java & Javac now.
Besides having AnyLogic installed, Java, & Alpyne is there any other prerequisites that are required?

**Log File:**
`[2021-11-12 07:54:42] INFO    @ com.anylogic.alpyne.AlpyneServer main 
                              | [PID=4704] Initializing @ 'C:\Users\test_user\Alpyne\examples\StockManagementGame\Exported\StockManagementGame' with arguments: {path=., port=8888, dev_mode=false}
[2021-11-12 07:54:42] FINER   @ ModelManager.<init>
                              | ENTRY {0}	.
[2021-11-12 07:54:42] FINER   @ BonsaiBrainLearningPlatform.<init>
                              | ENTRY {0}	{}
[2021-11-12 07:54:42] SEVERE  @ com.anylogic.alpyne.AlpyneServer main 
                              | Issue while attempting to run server: null
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at com.anylogic.alpyne.sim.api.JarReader.getRLExperiment(JarReader.java:40)
	at com.anylogic.alpyne.sim.ModelManager.<init>(ModelManager.java:42)
	at com.anylogic.alpyne.AlpyneServer.main(AlpyneServer.java:93)
Caused by: java.lang.NoClassDefFoundError: com/querydsl/sql/RelationalPath
	at com.anylogic.engine.ExperimentCustom.<init>(Unknown Source)
	at com.anylogic.engine.ExperimentReinforcementLearning.<init>(Unknown Source)
	at com.anylogic.engine.ExperimentReinforcementLearning.<init>(Unknown Source)
	at stock_management_game1.RLExperiment.<init>(RLExperiment.java:90)
	... 8 more
Caused by: java.lang.ClassNotFoundException: com.querydsl.sql.RelationalPath
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 12 more

[2021-11-12 07:54:42] WARNING @ com.anylogic.alpyne.AlpyneServer main 
                              | Alpyne terminated.
[2021-11-12 07:54:42] INFO    @ AlpyneServer.main
                              | [PID=4704] Initializing @ 'C:\Users\test_user\Alpyne\examples\StockManagementGame\Exported\StockManagementGame' with arguments: {path=., port=8888, dev_mode=false}
[2021-11-12 07:54:42] SEVERE  @ AlpyneServer.main
                              | Issue while attempting to run server: null
                                java.lang.reflect.InvocationTargetException
                                	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
                                	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
                                	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
                                	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
                                	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
                                	at com.anylogic.alpyne.sim.api.JarReader.getRLExperiment(JarReader.java:40)
                                	at com.anylogic.alpyne.sim.ModelManager.<init>(ModelManager.java:42)
                                	at com.anylogic.alpyne.AlpyneServer.main(AlpyneServer.java:93)
                                Caused by: java.lang.NoClassDefFoundError: com/querydsl/sql/RelationalPath
                                	at com.anylogic.engine.ExperimentCustom.<init>(Unknown Source)
                                	at com.anylogic.engine.ExperimentReinforcementLearning.<init>(Unknown Source)
                                	at com.anylogic.engine.ExperimentReinforcementLearning.<init>(Unknown Source)
                                	at stock_management_game1.RLExperiment.<init>(RLExperiment.java:90)
                                	... 8 more
                                Caused by: java.lang.ClassNotFoundException: com.querydsl.sql.RelationalPath
                                	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
                                	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
                                	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
                                	... 12 more
                                
[2021-11-12 07:54:42] WARNING @ AlpyneServer.main
                              | Alpyne terminated.
`

**Comment by t-wolfeadam** [comment]
To your question - no, there's not other prerequisites. The error shown in the log indices it couldn't find one of the model dependencies. I would make sure the "lib" folder wasn't removed from the exported model folder. I also noticed the port number was changed; this _technically_ shouldn't cause problems but, in case there's a more underlying issue, I would try with the default too. Let me know if these don't help (or if they do) and I can try suggesting a few other things.
